audit/base/base.go

```diff
@@ -0,0 +1,556 @@
+will not cause an error return as the connection is still valid. If using a timeout on the Context, make sure to
+use a new Context for each Send(). If using the parent Context, it is recommended to do context.WithoutCancel(ctx)
+to prevent accidental cancellation. It is better to let queue full errors happen and handle them than to use
+a context timeout.
+
+Errors that are Unrecoverable or context cancelled will be sent to Recover(). It is up to the caller
+to either handle these errors or ignore them. If you ignore them, you will lose audit records.
+*/
+func (c *Client) Send(ctx context.Context, msg msgs.Msg) error {
+	if msg.Type == msgs.ATUnknown || msg.Type > msgs.ControlPlane {
+		return Error{Err: fmt.Errorf("audit type (%d is invalid", msg.Type), Category: ErrValidation}
+	}
+
+	if err := msg.Record.Validate(); err != nil {
+		return Error{Err: err, Category: ErrValidation}
+	}
+
+	if msg.Record.Hook != nil {
+		var err error
+		msg.Record, err = msg.Record.Hook(msg.Record)
+		if err != nil {
+			return Error{Err: err, Category: ErrValidation}
+		}
+	}
+
+	// We always want to send the message unless the queue is full or the context times out
+	// before we get to send the message.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case c.sendCh <- streamData[SendMsg]{Data: SendMsg{Ctx: ctx, Msg: msg}}:
+	default:
+		return Error{Err: errors.New("queue is full"), Category: ErrQueueFull}
+	}
+
+	// If we had any errors previously, we need to return them.
+	if err := c.getErr(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Reset resets the connection to the audit server.
+// This can be used if the connection is in a bad state and needs to be reset.
+// This will cause an existing connection to be closed and reset internal state.
+// This is thread safe.
+func (c *Client) Reset(ctx context.Context, newConn conn.Audit) error {
+	if c == nil {
+		return Error{Err: fmt.Errorf("cannot call Reset on a base.Client that is nil"), Category: ErrConnection}
+	}
+	if newConn == nil {
+		return Error{Err: fmt.Errorf("cannot pass a conn.Audit that is nil"), Category: ErrConnection}
+	}
+
+	// If for some reason there is no error happening, we need to set one to prevents sends.
+	c.setErr(Error{Err: fmt.Errorf("audit client Reset() called, resetting connection"), Category: ErrConnection})
+
+	// If the connection is live, we need to close it.
+	c.close(ctx, false)
+	c.conn.Store(&newConn)
+
+	// Now let's bring everything back to life.
+	c.done = make(chan struct{})
+
+	c.closeOnceMu.Lock()
+	c.closeOnce = sync.Once{}
+	c.closeOnceMu.Unlock()
+
+	c.setErr(nil)
+	return nil
+}
+
+// Close closes the connection to the audit server.
+func (c *Client) Close(ctx context.Context) error {
+	c.wait()
+	return c.close(ctx, true)
+}
+
+// wait waits for the client to finish sending all messages. This is determined by the queue
+// being empty.
+func (c *Client) wait() {
+	for {
```

Comment:
can block forever, should this take a ctx?

Reply:
I don't think so????  So this is used in Close(), if your Close() does not empty the queue, then you have a problem.

The only use for Close() is before you exit your service. This is a rare occurrence. Otherwise you need to be using Reset().  Close() is also one of those things that gets ignored for errors all the time, especially at shut down.

If we want to be kind, we could do:

```go
func (c *Client) wait() {
         doneWaiting := make(chan struct{})
        go func() {
              defer close(doneWaiting)
              for {
                  if len(c.sendCh) == 0 {
		    	return
                  }
                  time.Sleep(100 * time.Millisecond)
              }
        }()
	for {
               select {
               case <-time.After(10 * time.Second):
                   log.Println("waiting for audit log base client to send all messages after a Close()")
               case <-doneWaiting:
                   return 
               }
	}
}
```

Reply:
Agree that passing a ctx here is wrong, and what you've suggested seems better.  I do think that this is probably worth doing over just blocking forever? 

Is there ever really an expected situation that this takes longer than 10s?

Reply:
Yes, if the connection is busted and the queue isn't emptying.